registry/storage/driver/s3-aws/s3.go

```diff
@@ -665,103 +665,104 @@ func (d *driver) Reader(ctx context.Context, path string, offset int64) (io.Read
-				Detail:     fmt.Errorf("append to zero-size path %s unsupported", path),
-			}
-		}
+func (d *driver) createMultipartUpload(ctx context.Context, key string) (*string, error) {
+	res, err := d.S3.CreateMultipartUploadWithContext(ctx, &s3.CreateMultipartUploadInput{
+		Bucket:               aws.String(d.Bucket),
+		Key:                  aws.String(key),
+		ContentType:          d.getContentType(),
+		ACL:                  d.getACL(),
+		ServerSideEncryption: d.getEncryptionMode(),
+		SSEKMSKeyId:          d.getSSEKMSKeyID(),
+		StorageClass:         d.getStorageClass(),
+	})
+	if err != nil {
+		return nil, err
+	}
+	return res.UploadId, nil
+}
 
-		var allParts []*s3.Part
-		for _, multi := range resp.Uploads {
-			if key != *multi.Key {
-				continue
-			}
+func (d *driver) inProgressUpload(ctx context.Context, path string) (uploadID *string, err error) {
+	var empty bool
+	if err := d.S3.ListMultipartUploadsPagesWithContext(ctx, &s3.ListMultipartUploadsInput{
+		Bucket: aws.String(d.Bucket),
+		Prefix: aws.String(d.s3Path(path)),
+	}, func(page *s3.ListMultipartUploadsOutput, lastPage bool) bool {
+		if len(page.Uploads) == 0 {
+			empty = true
+			return false
+		}
 
-			partsList, err := d.S3.ListPartsWithContext(ctx, &s3.ListPartsInput{
-				Bucket:   aws.String(d.Bucket),
-				Key:      aws.String(key),
-				UploadId: multi.UploadId,
-			})
-			if err != nil {
-				return nil, parseError(path, err)
-			}
-			allParts = append(allParts, partsList.Parts...)
-			for *partsList.IsTruncated {
-				partsList, err = d.S3.ListPartsWithContext(ctx, &s3.ListPartsInput{
-					Bucket:           aws.String(d.Bucket),
-					Key:              aws.String(key),
-					UploadId:         multi.UploadId,
-					PartNumberMarker: partsList.NextPartNumberMarker,
-				})
-				if err != nil {
-					return nil, parseError(path, err)
-				}
-				allParts = append(allParts, partsList.Parts...)
+		for _, upload := range page.Uploads {
+			if *upload.Key == d.s3Path(path) {
+				uploadID = upload.UploadId
+				return false
 			}
-			return d.newWriter(ctx, key, *multi.UploadId, allParts), nil
 		}
+		return true
+	}); err != nil {
+		return nil, fmt.Errorf("list multipart uploads pages: %w", err)
+	}
 
-		// resp.NextUploadIdMarker must have at least one element or we would have returned not found
-		listMultipartUploadsInput.UploadIdMarker = resp.NextUploadIdMarker
+	if !empty {
+		return uploadID, nil
+	}
 
-		// from the s3 api docs, IsTruncated "specifies whether (true) or not (false) all of the results were returned"
-		// if everything has been returned, break
-		if resp.IsTruncated == nil || !*resp.IsTruncated {
-			break
+	fi, err := d.Stat(ctx, path)
+	if err != nil {
+		return nil, parseError(path, err)
+	}
+
+	if fi.Size() > 0 {
+		return nil, storagedriver.Error{
+			DriverName: driverName,
+			Detail:     fmt.Errorf("append to non zero-size path %s unsupported", path),
 		}
 	}
-	return nil, storagedriver.PathNotFoundError{Path: path}
+	return nil, nil
```

Comment:
This rubs me the wrong way due to how I expect to consume return values from functions e.g. I would not expect the return value to be `nil` if the returned `error` is `nil` as well -- this often leads to very unexpected surprises for API consumers, especially when consuming unexported functions that often miss comments.

Reply:
Do you have any other suggestions? As long as it's documented then it's probably fine?

Reply:
Is there a reason why we no longer return `return nil, storagedriver.PathNotFoundError{Path: path}` like the original code did?

If there is no upload found in `path` then it would maybe make sense to return it and let the consumer of the API decide what to do in such situation (we'd obviously add a code comment explaining what's what)

Reply:
Hmm, yeah, sorry, I did have this in my original change but it somehow slipped when I switched it to using pages. I can fix that.

https://github.com/distribution/distribution/commit/d29ef835aa6b1f74e3fd6f02ffd579c28c1c8fcc

Reply:
Have resolved it in the latest change. I still don't see many other ways of expressing this condition, it's sort of nice for the zero value to have meaning? An extra return value to indicate whether it was found or not just feels redundant when it would essentially just be the value of `*uploadID != nil` anyway.

Reply:
I almost prefer the previous version I wrote without the page API just because it doesn't need to hack around the function signature `func(...) bool`. I kept it though because I thought it was better to not manage pagination manually.

Reply:
Yeah, we need to figure this out. Once this is "fixed" I'm ok with this PR getting merged, but returning two `nil`s is prone to very unexpected surprises which we should avoid.

Reply:
That doesn't mean we should carry them over, right?

Reply:
I don't see much reason not to? The std lib is usually a good reference for idiomatic Go.

Reply:
So these examples are kinda interesting, thanks for sharing them! I finally had some time to peek into them: a lot of them are returned when a slice is expected as a returned value which makes sense because slices are usually consumed by ranging/further slicing i.e. operations that wouldn't trigger `panic`s. Some of them are returned in internal packages, etc. But ok, maybe we do the same here since this is an unexported `func`. I think it's still worth adding a comment highlighting this to the "future" generations 🤷‍♂️

Reply:
Makes sense to me, thanks :) I'll add some comments.

Reply:
Done,

Reply:
So I'm wondering, if we should return something like this here instead of `nil, nil` @uhthomas. This will at least provide some signal to upstream API consumer of this function that the upload has not been found.

```suggestion
	return nil, storagedriver.Error{
		DriverName: driverName,
		Detail:     fmt.Errorf("no in-progress upload found for empty file at path %s", path),
	}
```

Reply:
The `nil` upload ID is being used as a valid value. There is only one consumer of this function and it does not fail or propagate the error, instead, it will create a new upload instead. The `Writer` function would need to explicitly handle this error instead, and would be messy without something a bit more structured?

So instead of checking `uploadID == nil` it would need to instead check if the error is a `storagedriver.Error` and then check if `err.Detail` looks like `""no in-progress upload found"`? I'm not sure it even should be an error if there is no in progress upload, it's expected.